### ISSUE TITLE
ProxyFix with x_port should skip standard 80 or 443 ports

### ISSUE THREAD

**Issue by n1ngu** [pallets/werkzeug#2839]
When a proxy is setting the X-Forwarded-Port header and a WSGI app is patched with `app = ProxyFix(app, x_port=1)`, a request to a standard port to the proxy adds the port to the hostname anyway.

Given the following minimal app
```
import wsgiref.simple_server
import werkzeug.middleware.proxy_fix

app = werkzeug.middleware.proxy_fix.ProxyFix(
    wsgiref.simple_server.demo_app,
    x_port=1,
)
```
when deployed behind an AWS ALB over standard HTTPS, visiting the website page will contain
```
HTTP_HOST = 'www.example.com:443'
SERVER_PORT = '443'
```

Because 443 is the standard port for the https protocol, the client would skip the `:443` fragment in the Host header. Likewise, no `:80` suffix would be added to the host header if http was the protocol.

A client would only put the port suffix in the Host header if it was using a non-standard port. I think ProxyFix's purpose is to patch the wsgi environ in a way that the downstream wsgi application can't tell whether it is being deployed behind an upstream proxy, so it should not add a standard port to the HTTP_HOST wsgi variable.

I may have misunderstood something, please tell me so.

Environment:

- Python version: 3.11
- Werkzeug version: 3.0.1

**Comment by n1ngu** [comment]
For the time being, I am using yet another downstream middleware as a workaround to strip that hostname suffix

```

def _ProxyFixFix(app):
    suffixes = {"https": ":443", "http": ":80"}

    def _proxy_fix_fix(environ, start_response):
        nosuffix = suffixes.get(environ["wsgi.url_scheme"])
        if nosuffix:
            environ["HTTP_HOST"] = environ["HTTP_HOST"].removesuffix(nosuffix)
        return app(environ, start_response)

    return _proxy_fix_fix
```

**Comment by n1ngu** [comment]
Re-reading myself, I made not a great job at explaining the issue. This would be a hell lot easier to draw on a paper while chatting. Let me try again.

If you request an application on a non-standard port, the http client would include the port in the host header
```
curl -v https://example:PORT
# ...
> GET / HTTP/1.1
> Host: example:PORT
> User-Agent: curl/8.5.0
> Accept: */*
# ...
```
but it wouldn't do so for a standard port. In a WSGI application, a standard port request would end up in the environment as
```py
{
    "HTTP_HOST" : "example",
    "SERVER_PORT": "443",
    "wsgi.url_scheme": "https",
}
```

Now if that is deployed behind a proxy that forwards the port (and, for example clarity, the protocol), the inner request to an internal-IP:8888 backend application would look like

```
GET / HTTP/1.1
Host: example
User-Agent: curl/8.5.0
X-Forwarded-Port: 443
X-Forwarded-Proto: https
```
which the wsgi server will turn into this environment
```py
{
    "HTTP_HOST" : "example",
    "SERVER_PORT": "8888",
    "HTTP_X_FORWARDED_PORT": "443",
    "HTTP_X_FORWARDED_PROTO": "https",
    "wsgi.url_scheme": "http",
}
```
which `ProxyFix(app, x_port=1, x_proto=1)` would translate into
```py
{
    "HTTP_HOST" : "example:443",
    "SERVER_PORT": "443",
    "wsgi.url_scheme": "https",
    "werkzeug.proxy_fix.orig": {
        "HTTP_HOST": "example",
        "wsgi.url_scheme": "http",
        "SERVER_PORT": "8888",
    }
}
```
whereas I think it should be
```py
{
    "HTTP_HOST" : "example",
    "SERVER_PORT": "443",
    "wsgi.url_scheme": "https",
    "werkzeug.proxy_fix.orig": {
        "wsgi.url_scheme": "http",
        "SERVER_PORT": "8888",
    }
}
```
just as if the application was deployed on https://example:443 without proxies.

The problem is the `HTTP_HOST` wsgi environ var will be used by downstream middleware and applications for multiple purposes: validating request origins, producing external URLs (for emails and whatnot), checking CORS policies...

And `"example" != "example:443"`, causing subtle malfunctions.

**Comment by n1ngu** [comment]
[`werkzeug.sansio.utils.get_host`](https://github.com/pallets/werkzeug/blob/d3dd65a27388fbd39d146caacf2563639ba622f0/src/werkzeug/sansio/utils.py#L85) does some handling for this case, si it could be argued that downstream middleware/applications should do likewise or use that very utility.

But I expected this to be done in ProxyFix itself from the start.

**Comment by davidism** [comment]
I don't think this needs to be changed. It's not more or less correct to add `:80` and `:443` after the host when using the standard proto/port pairs. Any tool interpreting the host value should be able to handle either form, since either is correct.

**Comment by n1ngu** [comment]
For the record, if your downstream WSGI stack could't be easily configured/patched to accept the equivalence of `hostname` and `hostname:{80,443}`, a consistent werkzeug-based workaround to this issue is this middlware:

```py
import werkzeug.middleware.proxy_fix
import werkzeug.sansio.utils

def ProxyFix(app, **kwargs):
    get_host = werkzeug.sansio.utils.get_host

    def _proxy_hostname_fix(environ, start_response):
        environ["HTTP_HOST"] = get_host(
            environ["wsgi.url_scheme"],
            environ["HTTP_HOST"],
        )
        return app(environ, start_response)

    return werkzeug.middleware.proxy_fix.ProxyFix(
        _proxy_hostname_fix,
        **kwargs,
    )
```
